Implemented bicubic downscaling

ruanformigoni opened the pull request:
This PR implements a bicubic filter to remedy issue #692.

## Update 10-mar-2023

### Nier Automata (2160p)

Current gamescope result:
![gamescope_2023-03-10_17-28-34 raw_encoded](https://user-images.githubusercontent.com/18180693/224428773-7534715f-d7e6-4e3c-851b-0cf4fa96ad9b.png)

Proposal:
![gamescope_2023-03-10_17-28-40 raw_encoded](https://user-images.githubusercontent.com/18180693/224428821-ec11c567-235c-4d74-927f-ff8f1b8822d2.png)

### Need for Speed ProStreet (2160p)

Current gamescope result:
![gamescope_2023-03-10_17-41-31 raw_encoded](https://user-images.githubusercontent.com/18180693/224429023-3cd7f145-c908-4c04-b505-e5d15dbfa27b.png)

Proposal:
![gamescope_2023-03-10_17-41-34 raw_encoded](https://user-images.githubusercontent.com/18180693/224429137-9c3d79e3-0f8d-45a4-8b2d-74461dfa1841.png)

### GTA IV  (4320p)

Current gamescope result:
![20230310165941_1](https://user-images.githubusercontent.com/18180693/224431068-b476544d-d76e-4afd-a979-bf3114545ea0.jpg)

Proposal:
![20230310165952_1](https://user-images.githubusercontent.com/18180693/224431077-aad4db91-57bf-4153-9625-df949aaab75d.jpg)

### Dark Souls II  (4320p)

Current gamescope result:
![screenshot-158](https://user-images.githubusercontent.com/18180693/230230074-8f1aacb9-c94c-4786-a246-18da26a11215.png)

Proposal:
![screenshot-159](https://user-images.githubusercontent.com/18180693/230230122-d561ed27-17ae-49d5-adce-75a371d0fb4c.png)

vic-bay commented:
Downsampling from 2160p to 1080p looks okay, but everything lower looks weird.
Here is comparison between 1440p and 1080p with FXAA. 1440p looks aliased and pixelated. 
![Screenshot_2023-01-12_15-30_12](https://user-images.githubusercontent.com/10389448/212083912-7d7e0ca7-f435-4f47-8a99-7b8edcb619ee.png)
![Screenshot_2023-01-12_15-31_56](https://user-images.githubusercontent.com/10389448/212083924-4ebf11b6-ecd6-49f3-b75c-2f77526e8052.png)

EndlesslyFlowering commented:
you should probably look into Catmull-Rom (Bicubic with b=0 and c=0.5) for downscaling rather than using normal bicubic

rKsanu2MMYvypWePtQWM commented:
I can confirm that resolutions that aren't double the monitor resolution look weird.

ruanformigoni commented:
Hey @EndlesslyFlowering , I've included the Catmull-Rom algorithm, thanks for bringing it to my attention. Thanks for the heads up @rKsanu2MMYvypWePtQWM, I've fixed this in the latest commits.

Here's the ranges of are shown below, the numbers represent the resolution multiplier from the monitor resolution and the render resolution, e.g., render at 2160p and display in 1080p == 2160/1080 == 2.

- Range of  [1,2) uses FSR Bilinear
- Range of [2,3) uses Catmull-Rom with a mean of 4 samples
- Range of [3,4) uses Catmull-Rom with a mean of 9 samples
- Range of (4,inf) uses Catmull-Rom with a mean of 16 samples

Displaysguy commented:
Definitely making really great progress. I think to cap it all off you should add the ability for user to adjust the bicubic parameters of B and C. Catmull-rom has ringing, which increases perceived sharpness but I want to be able to adjust it. 

Examples of different bicubic setups with the 2 parameters:

B-Spline(blur only): B1 C0

Hermite(blocking only): B0 C0

Mitchell(balanced minor amounts of blocking, aliasing, and ringing): B0.33 C0.33

Catmull-Rom(balance of ringing and blocking): B0 C0.5


Explanation of different scalers:
https://artoriuz.github.io/blog/mpv_upscaling.html
https://en.wikipedia.org/wiki/Mitchell%E2%80%93Netravali_filters

Image showing the effects of different bicubic parameters on scaled quality:
https://artoriuz.github.io/blog/images/mpv_upscaling/mitchell_survey.png

Also, if it's not too far out of scope, these scalers(especially the ringy ones like catmull-rom or lanzcos) work better when you convert gamma to sigmoid light before scaling, and back to the native gamma afterwards. 

Pic showing that with upscaling:
https://artoriuz.github.io/blog/images/mpv_upscaling/kanao/kanao_comparison.png

This applies also to the downscaling(your PR can be used as both an upscaler or downscaler, if it still works like the first iteration)

ruanformigoni commented:
@Displaysguy thanks a ton for the explanation and resources! I'll study how to implement this properly and look into the possible method for argument passing you mentioned [here](https://github.com/Plagman/gamescope/issues/692#issuecomment-1444992487).

specfreq commented:
I've had a chance to test this out a bit. [Here's some test screenshots.](https://imgsli.com/MTYyNTIx/1/2)

ruanformigoni commented:
Hi @specfreq , could you also include a screenshot with the downsampling algorithm disabled?

specfreq commented:
@ruanformigoni [Here's some new test screenshots](https://imgsli.com/MTYyNTky/2/4).
One thing I'm doing is changing the desktop resolution to match my target scaled resolution (1024x768 in this case) and forcing on vsync through the OS.

Launch options: `gamescope -D -f -W 1024 -H 768 -w 4096 -h 3072 -- %command%`

![1024x768](https://user-images.githubusercontent.com/122238666/225522369-6f71990a-42bd-4feb-ba2a-73c76571a331.png)
![Compositor](https://user-images.githubusercontent.com/122238666/225522373-bdc8b782-8208-4ad2-b2ed-e08dc900e1df.png)



Let me know if I don't have it set up correctly.

ruanformigoni commented:
Thanks for the followup @specfreq. I did the same as you (change the desktop resolution and downscale) and got way better results, [here is a comparison](https://imgsli.com/MTYyNjI4). Did you set the in-game render resolution to 4096x3072 in the resolution settings?

specfreq commented:
![20230316130642_1](https://user-images.githubusercontent.com/122238666/225740979-8ddc8863-5f84-41a1-bd3f-bc8faa99152c.jpg)
![20230316130057_1](https://user-images.githubusercontent.com/122238666/225740988-3f741941-d33f-4787-a3fc-ec70a108e714.jpg)

The resolution is set to 4096x3072


Here's my `git pull`: ![Screenshot_20230316_130902](https://user-images.githubusercontent.com/122238666/225741463-67e0ed67-e3f4-4a44-95d9-c2b2b8a07492.png)

ruanformigoni commented:
Are you able to toggle it with `super + k`?

ruanformigoni commented:
That should be fine, make sure that lines 176-179 are these:

```cpp
case KEY_K:
	g_wantedUpscaleFilter = (g_wantedUpscaleFilter == GamescopeUpscaleFilter::BICUBIC) ?
		GamescopeUpscaleFilter::LINEAR : GamescopeUpscaleFilter::BICUBIC;
	break;
```

In your launch command `gamescope -D -f -W 1024 -H 768 -w 4096 -h 3072 -- %command%`, do could you point it directly to where it is compiled?  E.g.: `~/Repositories/gamescope/build/gamescope -D -f -W 1024 -H 768 -w 4096 -h 3072 -- %command%`.

specfreq commented:
I just rebuilt it and it is working now.

specfreq commented:
Some test images and videos:
https://imgsli.com/MTY0NjU5

https://youtu.be/vQlEw4hNLwo
https://youtu.be/gmOfzJvFnwY
https://youtu.be/R2G3rUlbkLU

Is there any way to use this with integer scaling?

specfreq commented:
Here's a test video:
https://youtu.be/mbMsGK45p-U

Filoppi commented on src/shaders/cs_bicubic.comp:
```diff
@@ -0,0 +1,248 @@
+            ( -18 + 12 * B + 6 *C ) * ( f * f ) +
+            ( 6 - 2 * B ) ) / 6.0;
+    }
+    else if( f >= 1.0 && f < 2.0 )
```
I'm not too sure about it but, this branch seems unreachable?
You pass in f (x) as a fract of a pixel coordinates, thus it will never be beyond 0-1?

The original implementation, which I reimplemented here
https://github.com/dolphin-emu/dolphin/pull/11999/files#diff-368d3b7ce4f1890976b9f3b3d7f59df06cc906721d2dd882552a8503b3a40d7bR240
actually passed in fract*m (or n) here, effectively using all branches as it spanned from -1 to 2.
Thought to be honest I haven't 100% understood what the algorithm does so maybe my implementation is all wrong.

ruanformigoni replied:
Thanks for spotting this, I've updated the code and now it calculates the weights as described [here](https://gist.github.com/TheRealMJP/c83b8c0f46b63f3a88a5986f4fa982b1).

rKsanu2MMYvypWePtQWM commented:
any news? would love to see this finally merged...

Filoppi commented:
I highly suggest to take a look at this more advanced downsampling method:
https://github.com/dolphin-emu/dolphin/pull/11999/commits/ca93a5191fef40947a6fac935ce814fca8a2a295
It's heavier, but it looks damn fantastic when scaling from a relatively higher resolution. On older games, this would be perfect and very temporally stable.

Joshua-Ashton commented:
I don't understand why you are still running the whole RCAS pass, surely we don't want sharpening after doing downscaling?

What's the rationale behind doing FSR for bilinear at certain ranges? Does it look better? I have my doubts there.

I think it also makes sense to have a g_wantedDownscaleFilter and g_downscaleFilter. Shouldn't affect upscaling at all. I would be happy to default downscaling method.

Then do like:
```
				bool needsUpScaling = frameInfo.layers[0].scale.x < 0.999f && frameInfo.layers[0].scale.y < 0.999f;
                bool needsDownScaling = frameInfo.layers[0].scale.x > 1.001f && frameInfo.layers[0].scale.y > 1.001f;
                frameInfo.useBICUBICLayer0 = g_downscaleFilter == GamescopeDownscaleFilter::BICUBIC && needsDownScaling;
				frameInfo.useFSRLayer0 = g_upscaleFilter == GamescopeUpscaleFilter::FSR && needsUpScaling;
				frameInfo.useNISLayer0 = g_upscaleFilter == GamescopeUpscaleFilter::NIS && needsUpScaling;
```

ruanformigoni commented:
Hi @Joshua-Ashton , it was because I had issues running without it. I seem to have done so now, it uses `blit` instead, is that how it is supposed to work? I tried to only use a single `cmdBuffer` dispatch, but that didn't work well.

FSR (or its bilinear filter) looks better when the resolution is lower than 2X. I can try to experiment with bilinear in this range or leave as bicubic (since this PR is for a bicubic filter, using anything other would be mileading), to confirm this I've implemented another more computationally expensive bicubic method:

```cpp
vec4 textureBicubic( sampler2D textureSampler, vec2 texCoord )
{
	AF1 fWidth = AF1_AU1(c0.z);
	AF1 fHeight = AF1_AU1(c0.w);

    // Size of one texel 
	AF1 texelSizeX = ARcpF1(fWidth);
	AF1 texelSizeY = ARcpF1(fHeight);

    // Result
	vec4 nSum = vec4( 0.0, 0.0, 0.0, 0.0 );
	vec4 nDenom = vec4( 0.0, 0.0, 0.0, 0.0 );

    // Get the decimal part
	AF1 a = fract( texCoord.x * fWidth );
	AF1 b = fract( texCoord.y * fHeight );

    // B & C parameters
	float B = AF1_AU1(c2.x)/AF1_(100);
	float C = AF1_AU1(c2.y)/AF1_(100);

    // Bicubic calculation
	for( int m = -1; m <= 2; m++ )
	{
        for( int n = -1; n <= 2; n++)
        {
            vec4 vecData = texture(textureSampler, texCoord + vec2(texelSizeX * float( m ), texelSizeY * float( n )));
            float f1 = catMullRom( B, C, float( m ) - a );
            float f2 = catMullRom( B, C, -( float( n ) - b ) );
            vec4 vecCooef1 = vec4( f1,f1,f1,f1 );
            vec4 vecCoeef2 = vec4( f2, f2, f2, f2 );
            nSum = nSum + ( vecData * vecCoeef2 * vecCooef1  );
            nDenom = nDenom + (( vecCoeef2 * vecCooef1 ));
        }
	}
	return nSum / nDenom;
}

```

The results were almost identical to the current bicubic implementation, which is way faster. What do you think of an option such as `--force-needs-scaling` (not on this PR), this way one could use `fsr` or other filter independently for downscaling or upscaling.

Also, what do you thing about the passing of the B and C parameters? Is this `-B 0.0,0.5` and `--bicubic=0.0,0.5` ok or would you like to have only fixed values?

I have implemented the requested `g_downscaleFilter` and `g_wantedDownscaleFilter`

Joshua-Ashton commented:
I am so sorry, I didn't see this until now as it was posted when I was on holiday.

I think it's best to just keep downscale filter separate

ruanformigoni commented:
Hi @Joshua-Ashton , no problem :smile: . I'll try to improve the filter when possible, there is still [this issue](https://github.com/ValveSoftware/gamescope/issues/692#issuecomment-1576709102) pointed out by @DisplayTalk .

gapys-krzysztof commented:
Hey @ruanformigoni , any updates about this? :slightly_smiling_face: Seems it conflicts with the master branch

Also, is there going to be a way to use the bi**linear** filter for downscaling? For some people the current implementation of bicubic seeems a bit blurry in some cases, and those configurable B and C params (no clue what they do) seem to not affect the actual game on my end no matter what values I force them too, including small fractional and high integer.

One could technically hack FSR to also downscale (doesn't allow it by default), and it kinda works, but it wasn't designed for it and it's probably not a good idea.

dustContributor commented:
I'd love to have something like this on gamescope. It's really nice for older games because instead of having the GPU sit at 30% usage doing mostly nothing, you can push it a bit and get very nice image quality in return.

LethalManBoob commented:
Progress update?

Comptr-user commented:
Is there a way to get this to work?
I tried following the steps for building, but it refused to build. Using FSR for downsampling did build and run, however it was unusable.

TacoCake commented:
Hey, is this PR dead?

ruanformigoni commented:
Hello everyone, sorry for not going back to this sooner. I have managed to fix the remaining bugs on the bicubic filter.

## Bugs Fixed

1. The `B` and `C` parameters now work properly.
1. Image artifacts are no longer present
1. Upscaling now works correctly

## Recap on what this PR proposes

This PR introduces a bicubic filter for downscaling, as shown in the pictures
above, it produces images with higher quality than the current approach.

For the cli, two options are introduced `-D` and `--bicubic`, you can test the
bicubic filter for example, with:

`./build/src/gamescope -f --bicubic -w 2560 -h 1080 -- vkcube --width 5120 --height 2160`

Also, you can configure the `B` and `C` parameters for the [Mitchell–Netravali's](https://en.wikipedia.org/wiki/Mitchell%E2%80%93Netravali_filters)
filter, e.g.

`./build/src/gamescope -f --bicubic=1.0,0.0 -w 2560 -h 1080 -- vkcube --width 5120 --height 2160`


## Comparison with Iterative Bicubic

The current implementation is based on [this one](https://stackoverflow.com/questions/13501081/efficient-bicubic-filtering-code-in-glsl), it is more efficient computationally than other approaches I have tested. [Here](https://imgsli.com/MzAyNjU5/0/3) are some image quality comparisons from the current efficient approach with a heavier one, this is a upscaling test from 960x540 to 1920x1080.

## Other

I'll take some time tomorrow to rebase the changes to the newest commit. You can compile the current version with:
```
git submodule update --init
CXXFLAGS="-Wno-error" CFLAGS="-Wno-error" meson setup build
ninja -C build
build/gamescope -- <game>
```

rKsanu2MMYvypWePtQWM commented:
Thank you so much.

ruanformigoni committed:
Partial bicubic code adaptation

ruanformigoni committed:
Partial bicubic code adaptation [2]

ruanformigoni commented:
I have rebase the changes on another branch, unfortunately I am getting a black screen now, probably missed something that was introduced before the rebase. @misyltoad could you take a look at the [changes](https://github.com/ruanformigoni/gamescope/compare/master...issue-692-bicubic-rebase) and let me know what you think? Thanks!

LethalManBoob commented:
great to see this getting implemented, 4k linux gaming!!

dustContributor commented:
Thank you Ruan for continuing to work on this!

ruanformigoni committed:
Fix indentation

ruanformigoni committed:
Fix black screen on bicubic filter

ruanformigoni committed:
Use bicubic as an option of --filter

ruanformigoni commented:
Ok, I managed to fix the black screen issue @misyltoad, and also rebased the project to have no conflicts with master. I have modified how the arguments work to fit more what gamescope is doing right now with the `--filter` argument; so the usage now is `--filter=bicubic` or `-F bicubic`, with arguments it looks like `--filter=bicubic:0.3,0.3` or `-F bicubic:0.3,0.3`. My previous comment was updated with the full command line examples.

ruanformigoni committed:
Updated README with bicubic info

ruanformigoni committed:
Fix indentation in meson.build

matte-schwartz commented:
@ruanformigoni nice work on this!

something I've found while testing:

- I'm getting quite a few `[gamescope] [Error] xdg_backend: Compositor released us but we were not acquired. Oh no.` logs when using `gamescope -F bicubic -W 2560 -H 1080 -w 5120 -h 2160 -- vkcube` on Plasma 6 Wayland with the gamescope Wayland backend. this does not happen with the upscaling filters in gamescope, and it also does not seem to matter if Gamescope WSI is in use or not. I'm not seeing any obvious causes with `gamescopectl log_xdg_backend debug` but I haven't looked too far into it yet.

I will try to test some with the DRM backend sometime later today.

ruanformigoni commented:
Hi @matte-schwartz , thanks! I do not think this is related to the filter, since I also have this issue with the current implementation of FSR:
![logs](https://github.com/user-attachments/assets/b425dcea-7cf9-4d07-82ed-cb9cdea9b08a)

ruanformigoni committed:
Remove debug print

ruanformigoni committed:
Remove outdated comments

matte-schwartz commented:
Interesting, was not seeing the same on my rig with FSR although I was not testing in-game quite yet... just with vkcube. Could have just been a desktop compositor issue or something on my end I suppose. In general it's not a huge deal when that error appears but figured I'd make a note of it anyway.

gnusenpai commented:
I've found a strange bug. Enabling the bicubic filter destroys the gamma in some games, lowering contrast and making the colors washed out. It seems to affect all the Unreal Engine games I've tried.

arch1t3cht commented:
> Image showing the effects of different bicubic parameters on scaled quality:
https://artoriuz.github.io/blog/images/mpv_upscaling/mitchell_survey.png

(The image link is dead but it's still available in the wayback machine, or e.g. at https://upload.wikimedia.org/wikipedia/commons/thumb/6/6f/Mitchell-Netravali_artifacts.svg/1280px-Mitchell-Netravali_artifacts.svg.png )

Note that this graph (and the survey it's based on) is for *upscaling*. When downscaling, the parameters can have very different effects (depending on how the kernel is scaled with the downscaling ratio).

Artoriuz commented:
> Image showing the effects of different bicubic parameters on scaled quality: https://artoriuz.github.io/blog/images/mpv_upscaling/mitchell_survey.png

I realise I'm replying to a very old message, but I just wanted to point out that the information previously contained in the linked page has been somewhat moved to [this other page](https://artoriuz.github.io/blog/imagemagick_resampling.html). The content itself is different, but it covers the same topics.

LethalManBoob commented:
lets get this merged asap pls, linux gamers deserve hq super sampling :<

MCPO-Spartan-117 commented:
The help output needs to be updated, there isn't a Wayland keybind for this and Nvidia seems to have issues with this for whatever reason as @gnusenpai said.

Sp3ci3s8472 commented:
I build the project and installed the binaries to test it on my 7900XTX, sadly it does not seem to run.  The version I had before from my package manager (pacman) does seem to work (gamescope-3.15.14-1-x86_64).  Not sure what additional details you might need but I still have X11 instead of Wayland.
The command I used is:
gamescope -h 2160 -w 5160 -H 1440 -W 3440 -r 120 -- %command%

When I run it on my terminal:
`[gamescope] [Info]  console: gamescope version 6649fd8 (gcc 14.2.1)
No CAP_SYS_NICE, falling back to regular-priority compute and threads.
Performance will be affected.
[gamescope] [Info]  vulkan: selecting physical device 'AMD Radeon RX 7900 XTX (RADV NAVI31)': queue family 1 (general queue family 0)
[gamescope] [Info]  vulkan: physical device supports DRM format modifiers
[gamescope] [Info]  wlserver: [backend/headless/backend.c:67] Creating headless backend
[gamescope] [Info]  vulkan: supported DRM formats for sampling usage:
[gamescope] [Info]  vulkan: Creating Gamescope nested swapchain with format 44 and colorspace 0
gamescope: types/wlr_linux_dmabuf_v1.c:532: feedback_compile: Assertion `table_len > 0' failed.
Aborted (core dumped)
`

matte-schwartz commented:
@Sp3ci3s8472 that's unrelated, you need https://github.com/ValveSoftware/gamescope/pull/1548

Sp3ci3s8472 commented:
> @Sp3ci3s8472 that's unrelated, you need #1548

Thanks that indeed solved it!

doonv commented:
> @Sp3ci3s8472 that's unrelated, you need #1548

How do I apply that PR's changes when building this one with docker?

doonv commented:
> How do I apply that PR's changes when building this one with docker?

I found a scuffed way to do this, copy this PR's `src/rendervulkan.cpp`, modify it with #1548 patches, upload that to a pastebin website, then add a curl command in the `Dockerfile` so it downloads the modified file and puts it where its supposed to be. Cursed, but it works.

Boulotaur2024 commented:
Please merge this.
I'm planning to use it for upscaling actually. 
I'd love to have some gaussian upscaling to make things a little blurry in-game.
I like my games this way while most people want super sharp stuff, I know.
But not every one loves the "oil-painting look" of FSR, you know...

Reonu commented:
Very excited about this and hoping it gets merged soon

rKsanu2MMYvypWePtQWM commented:
Me too...

On 17 December 2024 01:13:01 EET, Reonu ***@***.***> wrote:
>Very excited about this and hoping it gets merged soon 
>
>-- 
>Reply to this email directly or view it on GitHub:
>https://github.com/ValveSoftware/gamescope/pull/740#issuecomment-2547085131
>You are receiving this because you were mentioned.
>
>Message ID: ***@***.***>

Eckoa commented:
> Please merge this. I'm planning to use it for upscaling actually. I'd love to have some gaussian upscaling to make things a little blurry in-game. I like my games this way while most people want super sharp stuff, I know. But not every one loves the "oil-painting look" of FSR, you know...

This is actually what im waiting for also, Bicubic does a much better job of preserving shapes in the image vs FSR/NIS even if the image is softer

R1chterScale commented:
In my testing this doesn't seem to work with HDR sadly, results in the image looking extremely washed out as though it was HDR image presented on an SDR display. Turning off the bicubic filter has HDR continuing to behave properly.

Natrox commented:
Throwing my hat in the ring. I would really like to see this merged. I have found, however, that setting a resolution around 2/3rd of the nested resolution in-game results in the eventual downscale looking better. 

For example, running with `-W 1920 -H 1080 -w 3840 -h 2160` (nested game resolution of 2160p, output resolution of 1080p) with an in-game resolution of 2560x1440 looks better than native 1920x1080 or the higher 3840x2160 (too sharp due to nearest neighbor filtering regardless of options.

My theory on why this is happening is flimsy, but:
I believe that when gamescope upscales from 1440p to 2160p, it does use some linear filtering. The subsequent nearest filtered downscale will pick the corner pixel of a 2x2 selection window, but since the 1440p to 2160p ratio is 1:1.5, it will pick a filtered pixel.

This is all speculation, I am mostly going off visual results.

I would really see this merged for one particular reason: **my TV can display 120Hz at 1080p**, but only 60Hz on higher resolutions. I really like to play games at 120Hz while also making use of super-sampling. Now that HDR has matured (I miss Auto HDR but inverse tonemapping is a good way to make use of the brightness boost) I can _almost_ leave Windows entirely.

ruanformigoni committed:
Fix missing keybinds

Calinou commented:
I'm not sure if it's just my eyes, but the screenshots in OP seem quite blurry with bicubic downsampling enabled. I'm aware the original bilinear filter will look too sharp (resulting in aliasing) especially at scale factors above 2x per axis, but I remember better results being possible.

We can probably take inspiration from emulators here, which is an area where rendering at very high resolutions is common, so there's a frequent need for high-quality downsampling. The state of the art in emulation seems to a few different techniques, such as [area downsampling](https://dolphin-emu.org/blog/2023/11/25/dolphin-progress-report-august-september-and-october-2023/#50-19939-implement-output-resampling-upscalingdownscaling-by-filoppi-and-samb) (supported in Dolphin) and [4xRGSS/NxAGSS](https://github.com/PCSX2/pcsx2/pull/7647) (supported in PCSX2).

Area downsampling seems more suited at very high downsampling factors, while 4xRGSS will be better for more moderate downsampling factors (between 1x and 2x per-axis), which I suppose will be the most common for PC gaming. 4xRGSS should give you a similar look as "16x MSAA" (which most graphics drivers implement as 4x SSAA + 4x MSAA), albeit without the MSAA part.

ruanformigoni commented:
Hi @Calinou, some screenshots are from the first version of the filter, and it has changed a lot and became less blurry, [here](https://imgsli.com/NDAxNjU4) is an up-to-date comparison between raw 4K and 4K to 1080p with the filter enabled. You can even tweak the B and C parameters through the CLI to make it even clearer, e.g., `--filter=bicubic:0.3,0.3`.

I think I'll create a Flatpak package that replaces the default `org.freedesktop.Platform.VulkanLayer.gamescope` one to ease the feedback on this. I can try to implement the mentioned filters in gamescope, if they provide better results without a large performance overhead, than it is a no-brainer to replace the current one. Also, feel free to send PR's to `issue-692-bicubic` and I'll test the changes.

R1chterScale commented:
@ruanformigoni I don't mean to overly pester but have you looked into/found the issues that occur with 10-bit colour?

ThomasEricB commented:
Valve merge this please.

HappyGoFishing commented:
> Enough is enough. If Valve is not going to merge any of the Pull Requests containing the downscaling code, might as well create a fork of my own: https://github.com/ThomasEricB/gamescope-lanczos-downscaling
> 
> Here is Crimson Desert being downscaled by Lanczos using the code: https://www.youtube.com/watch?v=jtYjmPNczGA

Thank you Thank you Thank you!!!!!!!!!
Do you have a paypal? I will happily donate to you.
